### PR TITLE
Make plugin Android Studio-compatible (Java 8 + optional dependency on externalSystem)

### PR DESCRIPTION
**Pull request by PawelLipski**
image

It wasn't working so far... since Android Studio still runs on Java 8, and we were building for Java 11.

I'm still not sure tho if Marketplace will classify the plugin as compatible with Android Studio:

image

They claim it isn't due to dependency on `externalSystem`... let's see if making this dep optional changes anything.

### CODE REVIEW

File: backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/BaseManagedBranchSnapshot.java [hunk 1]
```diff
@@ -40,7 +41,7 @@
    * parent data somewhere outside of this class (e.g. in {@link GitMacheteRepositorySnapshot}).
    */
   protected void setParentForChildren() {
-    for (var child : children) {
+    for (val child : children) {
```
Comment (PawelLipski) [inline backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/BaseManagedBranchSnapshot.java:44]: Changing from `var` to `lombok.val` is 90% of this PR; the other changes are about some APIs added to Java after 8

File: backendImpl/build.gradle [hunk 2]
```diff
@@ -1,3 +1,4 @@
+commonsIO()
```
Comment (PawelLipski) [inline backendImpl/build.gradle:1]: For a substitute of `InputStream#readAllBytes()`

File: backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepository.java [hunk 3]
```diff
@@ -77,8 +78,8 @@
   public IGitMacheteRepositorySnapshot createSnapshotForLayout(IBranchLayout branchLayout) throws GitMacheteException {
     LOG.startTimer().debug("Entering");
     try {
-      var aux = new CreateGitMacheteRepositoryAux(gitCoreRepository, statusHookExecutor, preRebaseHookExecutor);
-      var result = aux.createSnapshot(branchLayout);
+      val aux = new CreateGitMacheteRepositoryAux(gitCoreRepository, statusHookExecutor, preRebaseHookExecutor);
```
Comment (PawelLipski) [inline backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepository.java:81]: I could just use `lombok.var` and reduce the size of the diff... but this didn't work for some reason :/

File: branchLayoutImpl/src/main/java/com/virtuslab/branchlayout/impl/readwrite/BranchLayoutFileWriter.java [hunk 4]
```diff
@@ -50,10 +51,12 @@ public void write(Path path, IBranchLayout branchLayout, boolean backupOldFile)
       @NonNegative int level) {
 
     List<String> stringList = List.empty();
-    for (var entry : entries) {
-      var sb = new StringBuilder();
-      sb.append(String.valueOf(indentSpec.getIndentCharacter()).repeat(level * indentSpec.getIndentWidth()))
-          .append(entry.getName());
+    for (val entry : entries) {
+      val sb = new StringBuilder();
+      for (int i = 0; i < level * indentSpec.getIndentWidth(); i++) {
```
Comment (PawelLipski) [inline branchLayoutImpl/src/main/java/com/virtuslab/branchlayout/impl/readwrite/BranchLayoutFileWriter.java:56]: Meh

File: build.gradle [hunk 5]
```diff
@@ -140,7 +141,7 @@ allprojects {
 
 
   apply plugin: 'org.gradle.java-library'
-  sourceCompatibility = 11
+  sourceCompatibility = 8
```
Comment (PawelLipski) [inline build.gradle:144]: This is about what language features are available + what Java version is supposed to be recorded in classfiles (44 + 8 = 52.0, 44 + 11 = 55.0)

File: build.gradle [hunk 6]
```diff
@@ -159,6 +160,7 @@ allprojects {
         '-Werror', // Treat each compiler warning (esp. the ones coming from Checker Framework) as an error.
         '-Xlint:unchecked', // Warn of type-unsafe operations on generics.
       ]
+      options.release = 8
```
Comment (PawelLipski) [inline build.gradle:163]: This is about what Java APIs should be visible to our Java code... this is tricky b/c if you don't set that, then you can write Java 8 code calling Java 11 APIs (like `Path.of` or `InputStream#readAllBytes()`), which is bound to crash in runtime when running on Java 8-compatible JVM

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/FetchBackgroundable.java [hunk 7]
```diff
@@ -64,8 +65,8 @@ public void run(ProgressIndicator indicator) {
       // This method set a text under a progress bar (despite of docstring)
       indicator.setText(taskSubtitle);
     }
-    var fetchSupport = GitFetchSupport.fetchSupport(project);
-    var remote = remoteName.equals(LOCAL_REPOSITORY_NAME)
+    val fetchSupport = GitFetchSupport.fetchSupport(project);
+    GitRemote remote = remoteName.equals(LOCAL_REPOSITORY_NAME)
```
Comment (PawelLipski) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/FetchBackgroundable.java:69]: Sometimes `val` didn't work for some reasons I don't want to understand

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseRebaseBranchOntoParentAction.java [hunk 8]
```diff
@@ -159,56 +163,59 @@ private void doRebase(
...
-    var gitRebaseParameters = tryGitRebaseParameters.get();
+    val gitRebaseParameters = tryGitRebaseParameters.get();
     LOG.debug(() -> "Queuing machete-pre-rebase hooks background task for '${branchToRebase.getName()}' branch");
 
     new Task.Backgroundable(project, getString("action.GitMachete.BaseRebaseBranchOntoParentAction.hook.task-title")) {
       @Override
       @UIThreadUnsafe
       public void run(ProgressIndicator indicator) {
 
-        var wrapper = new Object() {
```
Comment (PawelLipski) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseRebaseBranchOntoParentAction.java:182]: One of the lesser known Java features... added somewhere b/w Java 9 and 11

File: logging/src/main/java/com/virtuslab/logger/EnhancedLambdaLogger.java [hunk 9]
```diff
@@ -16,8 +17,9 @@
   private final ThreadLocal<@Nullable Long> timerStartMillis;
 
   EnhancedLambdaLogger(Class<?> clazz) {
-    String category = clazz
-        .getPackageName()
```
Comment (PawelLipski) [inline logging/src/main/java/com/virtuslab/logger/EnhancedLambdaLogger.java:20]: Even that is missing from Java 8 🤦🏻

File: src/main/resources/META-INF/plugin.xml [hunk 10]
```diff
@@ -3,7 +3,7 @@
     <name>Git Machete</name>
     <vendor email="gitmachete@virtuslab.com" url="https://virtuslab.com">VirtusLab</vendor>
 
-    <depends>com.intellij.modules.externalSystem</depends>
+    <depends optional="true" config-file="com.virtuslab.git-machete-withExternalSystem.xml">com.intellij.modules.externalSystem</depends>
```
Comment (PawelLipski) [inline src/main/resources/META-INF/plugin.xml:6]: Let's see if that satisfies the Marketplace... I don't know if there's any way to know that upfront ☹️

File: src/main/resources/META-INF/plugin.xml [hunk 11]
```diff
@@ -50,18 +48,18 @@
...
-                <override-text place="GitMacheteContextMenu"/>
-            </action>
-
             <action id="GitMachete.RebaseSelectedBranchOntoParentAction"
                     class="com.virtuslab.gitmachete.frontend.actions.contextmenu.RebaseSelectedBranchOntoParentAction"
                     icon="MacheteIcons.REBASE">
                 <override-text place="GitMacheteContextMenu"/>
             </action>
 
+            <action id="GitMachete.OverrideForkPointOfSelectedBranchAction"
```
Comment (PawelLipski) [inline src/main/resources/META-INF/plugin.xml:57]: Unrel: moved to down to keep actions starting with `Checkout` next to each other (`Checkout` and `Checkout and Rebase onto Parent...`)

File: backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/hooks/PreRebaseHookExecutor.java [hunk 12]
```diff
@@ -66,28 +68,28 @@ public PreRebaseHookExecutor(Path mainDirectoryPath, Path gitDirectoryPath) {
...
 
       if (!completed) {
-        var message = "machete-pre-rebase hook (${hookFilePath}) for ${gitRebaseParameters} " +
+        val message = "machete-pre-rebase hook (${hookFilePath}) for ${gitRebaseParameters} " +
             "did not complete within ${EXECUTION_TIMEOUT_SECONDS} seconds; aborting the rebase";
         LOG.withTimeElapsed().error(message);
         throw new GitMacheteException(message
-            + (!strippedStdout.isBlank() ? NL + "stdout:" + NL + strippedStdout : "")
-            + (!strippedStderr.isBlank() ? NL + "stderr:" + NL + strippedStderr : ""));
+            + (!strippedStdout.trim().isEmpty() ? NL + "stdout:" + NL + strippedStdout : "")
```
Comment (mkondratek) [inline backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/hooks/PreRebaseHookExecutor.java:79]: :sad-pepe:

File: branchLayoutImpl/src/main/java/com/virtuslab/branchlayout/impl/readwrite/BranchLayoutFileWriter.java (hunk 4 above)
Comment (mkondratek) [reply branchLayoutImpl/src/main/java/com/virtuslab/branchlayout/impl/readwrite/BranchLayoutFileWriter.java:56]: What about extracting `level * indentSpec.getIndentWidth()` to the `{lombok.|}val`?

---

https://stackoverflow.com/questions/3881660/java-for-loop-evaluation

File: branchLayoutImpl/src/test/java/com/virtuslab/branchlayout/unit/BranchLayoutFileReaderTestSuite.java [hunk 13]
```diff
@@ -26,14 +28,14 @@
...
-  private final Path path = Path.of("");
+  private final Path path = Paths.get("");
 
   @SneakyThrows
   private BranchLayoutFileReader getBranchLayoutFileReaderForLines(List<String> linesToReturn, int indentWidth) {
     BranchLayoutFileReader reader = new BranchLayoutFileReader();
 
     PowerMockito.mockStatic(BranchLayoutFileUtils.class);
-    var indentSpec = new IndentSpec(/* indentCharacter */ ' ', indentWidth);
+    val indentSpec = new IndentSpec(/* indentCharacter */ ' ', indentWidth);
```
Comment (mkondratek) [inline branchLayoutImpl/src/test/java/com/virtuslab/branchlayout/unit/BranchLayoutFileReaderTestSuite.java:38]: `DEFAULT_INDENT_CHARACTER` here?

File: build.gradle [hunk 14]
```diff
@@ -338,6 +340,12 @@ subprojects {
       }
     }
 
+    commonsIO = { ->
+      dependencies {
+        implementation group: 'commons-io', name: 'commons-io', version: commonsIOVersion
```
Comment (mkondratek) [inline build.gradle:345]: another dependency :sad-pepe:

**Review by mkondratek — APPROVED** [review]
lgtm

**Pull request merged**